src/main/java/sorald/FileUtils.java

```diff
@@ -105,7 +107,24 @@ public static void writeJSON(File file, Object coreObj, Map<String, Object> addi
             throws IOException {
         // JSONObject's constructor recursively uses getter methods to produce a JSON object
         JSONObject jo = new JSONObject(coreObj);
-        additionalData.forEach(jo::put);
+
+        // Converting List and Arrays to JSONArray, and other objects to JSONObject
+        additionalData.entrySet().stream()
+                .map(
+                        e ->
+                                new AbstractMap.SimpleEntry(
+                                        e.getKey(),
+                                        (e.getValue() instanceof List
+                                                ? ((List) e.getValue()).toArray()
+                                                : e.getValue())))
+                .forEach(
+                        e ->
+                                jo.put(
+                                        (String) e.getKey(),
+                                        (e.getValue().getClass().isArray()
+                                                ? new JSONArray(e.getValue())
+                                                : new JSONObject(e.getValue()))));
```

Comment:
This is just a nit-pick, but the reason you need to cast to `String` here is because you're using a raw collection. I also realized there's a bug: you should check for an instance of a `Collection`, not `List` (it could be e.g. a set or something like that as well, which should be a JSONArray).

```suggestion
        additionalData.entrySet().stream()
                .map(
                        e ->
                                new AbstractMap.SimpleEntry<>(
                                        e.getKey(),
                                        (e.getValue() instanceof Collection
                                                ? ((Collection<?>) e.getValue()).toArray()
                                                : e.getValue())))
                .forEach(
                        e ->
                                jo.put(
                                        e.getKey(),
                                        (e.getValue().getClass().isArray()
                                                ? new JSONArray(e.getValue())
                                                : new JSONObject(e.getValue()))));
```

Raw types really should be avoided if at all possible, they force casting too much.

Anyway, now that I'm looking at this, I think it would be much nicer to just define a helper method like so:

```java
    /**
     * Convert any collection or array into a JSONArray, and anything else into a JSONObject.
     */
    private static Object toJSONArrayOrObject(Object obj) {
        if (obj instanceof Collection) {
            return new JSONArray(((Collection<?>) obj).toArray());
        } else if (obj.getClass().isArray()) {
            return new JSONArray(obj);
        } else {
            return new JSONObject(obj);
        }
    }
```

And then make the whole thing be this:
```suggestion
        additionalData.forEach((k, v) -> jo.put(k, toJSONArrayOrObject(v));
```

Streams often get a bit unreadable if you define everything inline, so helpers methods can make it a lot more clear what's going on. But you decide what you wanna do here, I just find discussing this stuff to be really fun :)